`Iris`: Add chat button & widget, and a pop-up window

asliayk opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [X] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [X] Language: We followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] We chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [X] **Important**: We implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [X] We followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), we specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [X] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
A chat widget is needed to chat with our chatbot(Iris), and a button is needed to open the chat widget.
A pop-up window is needed to request chat permission from the user.

### Description
<!-- Describe your changes in detail -->
- Added a chat button on the right bottom of the exercise page.
- Added a chat widget to chat with Iris. It is opened when the user clicks on the chat button.
- Added a pop-up window to get the permission of the user. It is opened when the user clicks on the chat button. If the user gives the permission, it won't be opened next time, instead the chat widget will be opened.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Student
- 1 Programming Exercise

1. Log in to Artemis.
2. Navigate to a course page to see its exercises.
3. Select a Programming Exercise and navigate to its page.
4. Chat button is on the right bottom of the page.
5. Hover over it to see the helping tooltip. ("Start chatting with Iris!")
6. Click on it too see the pop-up window. 
7. Decline to observe that no window is opened.
8. Accept the permission to see that the windows is opened.
9. Check if the chatbot show the last message at the bottom by sending a lot of messages.
10. Observe the pop-up & chatbot visualisation in the dark mode. (Dialog Box will not be updated for Dark Mode because it will be deprecated in the next PR.)
11. Observe the pop-up & chatbot texts in the German language. (The first message will stay in English because it will be changed later.)


#### Performance Review
- [X] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Test Coverage

### Screenshots
![image](https://github.com/ls1intum/Artemis/assets/56061820/dd1e198e-7454-45de-bb41-c32f2a56aa9d)
![image](https://github.com/ls1intum/Artemis/assets/56061820/6f02d1d4-e6fe-4206-9cd0-0391915bbb73)

Hialus commented on src/main/webapp/app/overview/exercise-details/course-exercise-details.component.html:
```diff
@@ -251,4 +251,5 @@ <h3>{{ 'artemisApp.courseOverview.exerciseDetails.allResults' | artemisTranslate
         </div>
     </div>
     <jhi-lti-initializer></jhi-lti-initializer>
+    <jhi-exercise-chatbot></jhi-exercise-chatbot>
```
If you add the button here, you'll have to make sure that it is only displayed when the exercise is a programming exercise, since this is the only type of excercise we want to support for now.

Hialus replied:
Why can't you create a programming exercise? 
For this exact reason it is essentail that you have a working local setup.

eceeeren replied:
We tried to install it with the Intructor for 1 hour extra and it kept on giving Internal Server Error. I think there is not more I can do at this point. Maybe Asli would be able to create for herself. But then you will have to wait for the basic UI representation.

eceeeren replied:
@eceeeren
[Chatbot made available only for programming](https://github.com/ls1intum/Artemis/pull/6584/commits/976390a9cc8ad1ca9032e89e890d82203993f518)

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.scss:
```diff
@@ -0,0 +1,82 @@
+    padding: 10px;
+    border-radius: 10px;
+    margin-bottom: 10px;
+    //align-self: flex-end;
```
Please remove this if it is no longer needed.

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.html:
```diff
@@ -0,0 +1,31 @@
+
+        <!-- input field section -->
+        <div class="chat-input">
+            <input [(ngModel)]="newMessage" type="text" (keyup.enter)="onSend()" placeholder="Write a message..." />
```
Also use i18n for this

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts:
```diff
@@ -0,0 +1,35 @@
+})
+export class ExerciseChatWidgetComponent {
+    messages: string[] = [];
+    IrisMessages: string[] = ['Hey! How can I help you?'];
```
Make sure to use proper code styling.
```suggestion
    irisMessages: string[] = ['Hey! How can I help you?'];
```

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatbot.component.ts:
```diff
@@ -0,0 +1,35 @@
+    openDialog() {
+        const dialogRef = this.dialog.open(ChatbotPopupComponent, {
+            data: {
+                name: 'User',
```
Please use the actual first name of the user here.

asliayk replied:
I changed the code, the first name of the user is visible on the pop-up window now.

Hialus reviewed (changes requested):
Looked over the code a bit and left some first remarks.
Also I haven't tested it yet, but from the code it seems like you do not support dark mode. Please fix this, as this is also part of the PR checklist.

eceeeren commented on src/main/webapp/app/overview/exercise-chatbot/chatbot-popup/chatbot-popup.component.html:
```diff
@@ -0,0 +1,7 @@
+<h2 mat-dialog-title>Welcome {{ firstName }}!</h2>
+<mat-dialog-content> After you accept the regulations, your chat with IrisBot will start. Please decline if you do not wish to chat.</mat-dialog-content>
```
Added. Please recommend other strings if the use of the language is wrong

SabrinaGlatz reviewed (commented):
I tested it manually on a test server. It works fine, but for a second, I see the pop-up window somewhere else on the screen (see screenshot)
<img width="893" alt="image" src="https://github.com/ls1intum/Artemis/assets/125522114/a443562c-1deb-4f5c-a9a6-b75c69fd517e">

asliayk commented:
the bugs denoted by @SabrinaGlatz has been fixed

asliayk commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatbot.component.html:
```diff
@@ -0,0 +1,8 @@
+<div class="chatbot-button">
+    <button class="btn btn-primary btn-md rounded-circle" *ngIf="chatAccepted !== 'true'" (click)="openDialog()">
```
I added a tooltip, and refactored the code to have only one button in html.
I also changed the button shape, it is circle now.

undernagruzez commented on src/main/webapp/app/overview/exercise-chatbot/chatbot-popup/chatbot-popup.component.ts:
```diff
@@ -0,0 +1,24 @@
+
+    ngOnInit() {
+        this.authStateSubscription = this.accountService
+            .getAuthenticationState()
```
Not sure, but probably get authentication state is used to keep track of login/logout events. What you actually need is

```suggestion
            this.accountService.identity().then((user: User) => {
            this.user = user!;
        });
```

asliayk replied:
I updated it, now works with your suggestion

undernagruzez commented on src/main/webapp/i18n/de/student-dashboard.json:
```diff
@@ -197,7 +197,13 @@
-                "complaintPossibleTooltip": "Nach Erhalt einer Bewertung wirst du {{days}} Tage zum Einreichen einer Beschwerde haben."
+                "complaintPossibleTooltip": "Nach Erhalt einer Bewertung wirst du {{days}} Tage zum Einreichen einer Beschwerde haben.",
+                "chatWelcome": "Willkommen",
+                "chatPopUpMessage": "Nachdem Sie die Bestimmungen akzeptiert haben, beginnt Ihr Chat mit IrisBot. Bitte lehnen Sie ab, wenn Sie nicht chatten möchten.",
```
Imo this message should also inform a user about future transmission of their data to third-party servers.

eceeeren replied:
Done

undernagruzez commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatbot.component.ts:
```diff
@@ -0,0 +1,62 @@
+    faCommentDots = faCommentDots;
+
+    constructor(private dialog: MatDialog, private router: Router) {
+        this.router.events.subscribe((event) => {
```
Is there a specific reason why you are using router events subscription here? Won't the check in onNgDestroy solve the problem?

undernagruzez replied:
Also, you should unsubscribe in onNgDestroy anyways.

asliayk replied:
you are right, this is a shorter solution. I updated it.

undernagruzez reviewed (approved):
Minor code changes are necessary but can be merged now.

SabrinaGlatz reviewed (commented):
Looks good. Thanks for fixing the previous issues. I just recognized a small issue. On Tuesday, we discussed that the student can still interact with the background after opening the pop-up. But when i open the chat, I can't scroll down to read the rest of the problem statement.

dakennguyen reviewed (approved):
tested on ts1, it works well for me. But I have a suggestion that can be improved in another PR: maybe use vh/vw instead of px for width and height. The dialog and icon are a bit small for my screen
![Screenshot 2023-05-20 at 03 24 44](https://github.com/ls1intum/Artemis/assets/18320285/07e5a702-2ec7-4e80-bd61-84d98fc992f7)
Thank you for the amazing work!

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatbot.component.scss:
```diff
@@ -0,0 +1,13 @@
+    z-index: 9999;
+}
+
+.custom-circle-button {
```
Why don't you simply use the existing `btn-circle` for this?

asliayk replied:
updated

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatbot.component.ts:
```diff
@@ -0,0 +1,59 @@
+
+    constructor(private dialog: MatDialog) {}
+
+    ngOnDestroy() {
```
You've got a valid client style warning here. Please fix this.
> Lifecycle interface 'OnDestroy' should be implemented for method 'ngOnDestroy'. (https://angular.io/styleguide#style-09-01)

asliayk replied:
fixed

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.html:
```diff
@@ -0,0 +1,36 @@
+        <!-- client -->
+        <div class="client">
+            <div class="chat-header">
+                <button class="header-icon">
```
What is this button used for?

asliayk replied:
this button will be used for resizing the widget (or can open it as a new window)

Hialus replied:
Alright. 👍🏻

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.html:
```diff
@@ -0,0 +1,36 @@
+                (keyup.enter)="onSend()"
+                placeholder="{{ 'artemisApp.courseOverview.exerciseDetails.chatInputMessage' | artemisTranslate }}"
+            />
+            <button (click)="onSend()" class="send-btn">
```
Any reason why this doesn't use the standard Bootstrap button classes?

asliayk replied:
now uses btn-circle

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.scss:
```diff
@@ -0,0 +1,81 @@
+    background-color: #212529;
+    color: white;
+    font-weight: bold;
+    font-family: Arial;
```
Another client style warning. Please address this:
> Unexpected missing generic font family (font-family-no-missing-generic-family-keyword)

Also do you really need to use a special font family declaration here? At least for me it doesn't make much of a difference whether this is Arial or the default font.

asliayk replied:
fixed

Hialus commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts:
```diff
@@ -0,0 +1,45 @@
+import { Component, ElementRef, EventEmitter, Output, ViewChild } from '@angular/core';
```
Client style warnings for unused imports. Please fix

asliayk replied:
fixed

Hialus commented on src/main/webapp/i18n/de/student-dashboard.json:
```diff
@@ -197,7 +197,13 @@
                 "complaintPossible": "Beschwerde möglich:",
-                "complaintPossibleTooltip": "Nach Erhalt einer Bewertung wirst du {{days}} Tage zum Einreichen einer Beschwerde haben."
+                "complaintPossibleTooltip": "Nach Erhalt einer Bewertung wirst du {{days}} Tage zum Einreichen einer Beschwerde haben.",
+                "chatWelcome": "Willkommen",
```
I think it may make sense to create a separate file for these translation strings. 
Though at the very least you should indicate what is meant by "chat" here, by e.g. putting it into an "iris" subgroup

eceeeren replied:
Done

Hialus reviewed (changes requested):
Tested on TS9.
Some Issues and possible improvements I encountered:
EDIT: It seems like your build is currently failing. Due to this an old version may have been deployed to TS9, so it is possible some of these issues are already resolved. Please fix your build though.
- The regulations dialog is not dark mode compatible: <img width="1285" alt="grafik" src="https://github.com/ls1intum/Artemis/assets/18218285/b1dd3fc1-e0f3-432c-9e42-6dfe4bdc0ed4">
- The send button is not circular, but pill shaped
- When hovering over or clicking on the send button it does not reflect this by changing color, like other buttons (e.g. the open chat button does this)
- When the chat window is open it would be nice if pressing the chat button again would close it.
- The close button is extremely small. For me it is only 16x30 on a 4k screen.
- The chat button is positioned weirdly imo. It is too hight up, especially when compared to the hint button: <img width="571" alt="grafik" src="https://github.com/ls1intum/Artemis/assets/18218285/2fecdaad-ab4a-45fa-a8ab-f030ef09b477">

I also left some comments for the code.

asliayk commented:
> Tested on TS9. Some Issues and possible improvements I encountered: EDIT: It seems like your build is currently failing. Due to this an old version may have been deployed to TS9, so it is possible some of these issues are already resolved. Please fix your build though.
> 
> * The regulations dialog is not dark mode compatible: <img alt="grafik" width="1285" src="https://user-images.githubusercontent.com/18218285/239695320-b1dd3fc1-e0f3-432c-9e42-6dfe4bdc0ed4.png">
> * The send button is not circular, but pill shaped
> * When hovering over or clicking on the send button it does not reflect this by changing color, like other buttons (e.g. the open chat button does this)
> * When the chat window is open it would be nice if pressing the chat button again would close it.
> * The close button is extremely small. For me it is only 16x30 on a 4k screen.
> * The chat button is positioned weirdly imo. It is too hight up, especially when compared to the hint button: <img alt="grafik" width="571" src="https://user-images.githubusercontent.com/18218285/239695942-2fecdaad-ab4a-45fa-a8ab-f030ef09b477.png">
> 
> I also left some comments for the code.

We made the following updates:
- The regulations dialog does not need to be dark mode compatible, because it will be removed in the next version of it. (instead of popup window, chat widget will ask for permission as a first message, as we have discussed in our general meeting last week)
- The send button is circular now.
- The color of chatbot button changes now when hovering over it.
- We made the chat widget close when the chatbot button is pressed.
- We enlarged the close button, and expand button (the upper left button of the chat widget).
- We changed the chat button position, I think it is better now. (We have already updated the code so that chatbot only appears in programming exercises, but we can still test it only in other types of exercises, that's why the position may not be optimal.)

RY997 reviewed (commented):
Tried to tested on ts1 and ts2, but encountered "Internal server error" many times, have no idea if this is an isolated case that happened to me or something broke after the changes

eceeeren commented:
> Looks good. Thanks for fixing the previous issues. I just recognized a small issue. On Tuesday, we discussed that the student can still interact with the background after opening the pop-up. But when i open the chat, I can't scroll down to read the rest of the problem statement.

Scrolling down problem should be fixed now.

asliayk commented:
> Tried to tested on ts1 and ts2, but encountered "Internal server error" many times, have no idea if this is an isolated case that happened to me or something broke after the changes

fixed now

SabrinaGlatz reviewed (dismissed):
Tested it again on TS2. Everything works as expected and all previous issues I mentioned are fixed. There are some chat widget specific things for the future I would like to discuss with you, but they are not relevant for merging this PR :)

Hialus reviewed (approved):
Code looks okay to me now, though it would be great if you could get a client expert from the Artemis dev team to look over it as well.
UI also looks alright now, though there is one point for future improvement: Viewing it on a small screen (mobile phone):
The chat button overlaps with the help button:
![IMG_2382](https://github.com/ls1intum/Artemis/assets/18218285/6850c222-3bef-4dc1-a3b0-0c317494e51b)
The chat window is not really position very well:
![IMG_2384](https://github.com/ls1intum/Artemis/assets/18218285/5a06b602-263b-4bce-b9d2-cc08001b6eee)

Again this is something that you can do in a follow up PR and does not have to happen now.

RY997 reviewed (dismissed):
Approve!

undernagruzez reviewed (approved):
lgtm

RY997 reviewed (approved):
re-approve

eceeeren reviewed (dismissed):
Works as expected.

bassner commented on src/test/javascript/spec/component/iris/exercise-chatbot.component.spec.ts:
```diff
@@ -0,0 +1,86 @@
+                { provide: MatDialog, useValue: mockDialog },
+                { provide: Overlay, useValue: mockOverlay },
+            ],
+        }).compileComponents();
+    });
+
+    beforeEach(() => {
+        fixture = TestBed.createComponent(ExerciseChatbotComponent);
```
You don't need the second `beforeEach` block. Just add it after then compileComponents call like this: 

```ts
compileComponents().then(() => {
    // Your code
});

eceeeren replied:
Done

bassner commented on src/test/javascript/spec/component/iris/exercise-chatbot.component.spec.ts:
```diff
@@ -0,0 +1,86 @@
+        fixture.detectChanges();
+    });
+
+    it('should create the component', () => {
+        expect(component).toBeTruthy();
+    });
```
This test is useless, let's remove it entirely.

eceeeren replied:
Done

bassner commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatbot.component.ts:
```diff
@@ -0,0 +1,66 @@
+    styleUrls: ['./exercise-chatbot.component.scss'],
+})
+export class ExerciseChatbotComponent implements OnDestroy {
+    public chatAccepted = 'false';
```
Why is this a string and not a boolean?

eceeeren replied:
Must be a misunderstanding, changed

eceeeren replied:
Sorry, this was done to get the value of the button after it is clicked, which is a string. But changed to be controlled in a different way and made boolean instead

bassner commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.scss:
```diff
@@ -0,0 +1,85 @@
+    margin-bottom: 10px;
+}
+
+.my-chat {
+    width: fit-content;
+    margin-left: auto;
+    background-color: #b7dfb9;
+    color: dimgrey;
+    padding: 10px;
+    border-radius: 10px;
+    margin-bottom: 10px;
+}
+
+.chat-input {
+    display: flex;
+    align-items: center;
+    justify-content: space-between;
+    padding: 10px;
+    background-color: #f2f2f2;
+}
+
+.chat-input input {
+    flex: 1;
+    margin-right: 10px;
+    padding: 5px 15px;
+    border: 1px solid lightgray;
+    border-radius: 20px;
+}
+
+.btn-circle {
+    width: 40px;
+    height: 40px;
+    border-radius: 50%;
+    background-color: #4caf50;
+    border: none;
+}
+
+.btn-circle:hover,
+.btn-circle:active {
+    background-color: #3e913e;
```
Please do not hardcode colors. Use predefined default colors instead; only specify your own colors if strictly necessary. Please follow the guidelines on client theming: https://docs.artemis.cit.tum.de/dev/guidelines/client-design/

eceeeren replied:
Done

bassner commented on src/main/webapp/app/overview/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts:
```diff
@@ -0,0 +1,45 @@
+    @ViewChild('chatWidget') chatWidget!: ElementRef;
+    @ViewChild('chatBody') chatBody!: ElementRef;
+    messages: string[] = [];
+    irisMessages: string[] = ['Hey! How can I help you?'];
```
Should we translate this initial message or is this only something temporary?

eceeeren replied:
Yes, this part will be changed in the next PR

bassner commented on src/main/webapp/content/scss/themes/theme-dark.scss:
```diff
@@ -117,6 +117,47 @@ html {
+    background-color: rgb(24, 26, 24) !important;
+}
+
+.chat-header {
+    background-color: var(--artemis-dark) !important;
+}
+
+.chat-body {
+    border: 1px solid rgb(24, 26, 24) !important;
+}
+
+.chat-body::-webkit-scrollbar {
+    background-color: #262626 !important;
+    width: 10px !important;
+}
+
+.chat-body::-webkit-scrollbar-thumb {
+    background-color: #363735 !important;
+}
+
+.client-chat {
+    color: white !important;
+    background-color: #363735 !important;
+}
+
+.my-chat {
+    color: white !important;
+    background-color: #357a38 !important;
+}
+
+.chat-input {
+    background-color: #262626 !important;
+}
+
+.chat-input input {
+    color: white !important;
+    background-color: #363735 !important;
+    border: 1px solid dimgray !important;
+}
+
```
Please do not hardcode colors. Use predefined default colors instead; only specify your own colors if strictly necessary. Please follow the guidelines on client theming: https://docs.artemis.cit.tum.de/dev/guidelines/client-design/

Why do you specify component specific styles in the global theme style sheet at all?
If you want to have different colors for both themes, please use CSS variables. There is a section on how to specify colors for each theme in the linked documentation above, so please give the entire page a good read.

I think this entire section should be removed from this file.

eceeeren replied:
Done

bassner commented on src/main/webapp/i18n/de/exerciseChatbot.json:
```diff
@@ -0,0 +1,12 @@
+    "artemisApp": {
+        "exerciseChatbot": {
+            "welcome": "Willkommen",
+            "popUpMessage": "Nachdem Sie die Bestimmungen akzeptiert haben, beginnt Ihr Chat mit IrisBot und Ihre Daten werden an die Server Dritter weitergegeben. Bitte lehnen Sie ab, wenn Sie nicht chatten möchten.",
+            "inputMessage": "Schreiben Sie eine Nachricht...",
+            "accept": "Akzeptieren",
+            "decline": "Ablehnen",
+            "buttonTooltip": "Beginnen Sie mit Iris zu chatten!"
```
Usually we use "Du" instead of "Sie" in German texts.

eceeeren replied:
Done

bassner commented on src/main/webapp/i18n/en/exerciseChatbot.json:
```diff
@@ -0,0 +1,12 @@
+    "artemisApp": {
+        "exerciseChatbot": {
+            "welcome": "Welcome",
+            "popUpMessage": "After you accept the regulations, your chat with IrisBot will start and your data will be shared with the third-party servers. Please decline if you do not wish to chat.",
```
Can we be more specific? We should say something like

```
By choosing to continue, you agree that your interactions with IrisBot will be processed by Microsoft and OpenAI, with data transfer occurring outside of our university data center. If you do not agree with these terms, please select 'Decline'. To acknowledge this and begin your chat with IrisBot, press 'Accept'.
```

eceeeren replied:
Done

bassner commented on src/test/javascript/spec/component/iris/exercise-chat-widget.component.spec.ts:
```diff
@@ -0,0 +1,46 @@
+        await TestBed.configureTestingModule({
+            imports: [FormsModule, FontAwesomeModule, MatDialogModule],
+            declarations: [ExerciseChatWidgetComponent, ChatbotPopupComponent, MockPipe(ArtemisTranslatePipe)],
+        }).compileComponents();
+    });
+
+    beforeEach(() => {
```
You don't need the second `beforeEach` block. Just add it after then compileComponents call like this: 

```ts
compileComponents().then(() => {
    // Your code
});

eceeeren replied:
Done

bassner commented on src/test/javascript/spec/component/iris/exercise-chat-widget.component.spec.ts:
```diff
@@ -0,0 +1,46 @@
+        fixture.detectChanges();
+    });
+
+    it('should create the component', () => {
+        expect(component).toBeTruthy();
+    });
```
This test is useless, let's remove it entirely.

eceeeren replied:
Done

bassner reviewed (changes requested):
Hi! I left a few comments below; please have a look at them.
